extensions/ql-vscode/test/vscode-tests/minimal-workspace/queries-panel/query-discovery.test.ts

```diff
@@ -1,203 +1,196 @@
+      makeTestFile(join(workspacePath, "dir1", "query2.ql"));
+      makeTestFile(join(workspacePath, "dir2", "query3.ql"));
+      makeTestFile(join(workspacePath, "query4.ql"));
+
+      await discovery.initialRefresh();
+
+      expect(discovery.buildQueryTree()).toEqual([
+        new FileTreeDirectory(workspacePath, "workspace", env, [
+          new FileTreeDirectory(join(workspacePath, "dir1"), "dir1", env, [
+            new FileTreeLeaf(
+              join(workspacePath, "dir1", "query1.ql"),
+              "query1.ql",
+              "java",
+            ),
+            new FileTreeLeaf(
+              join(workspacePath, "dir1", "query2.ql"),
+              "query2.ql",
+              "java",
+            ),
+          ]),
+          new FileTreeDirectory(join(workspacePath, "dir2"), "dir2", env, [
+            new FileTreeLeaf(
+              join(workspacePath, "dir2", "query3.ql"),
+              "query3.ql",
+              "java",
+            ),
+          ]),
+          new FileTreeLeaf(
+            join(workspacePath, "query4.ql"),
+            "query4.ql",
+            "java",
+          ),
+        ]),
+      ]);
+    });
 
-      await discovery.waitForCurrentRefresh();
+    it("should collapse directories containing only a single element", async () => {
+      makeTestFile(join(workspacePath, "query1.ql"));
+      makeTestFile(join(workspacePath, "foo", "bar", "baz", "query2.ql"));
+
+      await discovery.initialRefresh();
+
+      expect(discovery.buildQueryTree()).toEqual([
+        new FileTreeDirectory(workspacePath, "workspace", env, [
+          new FileTreeDirectory(
+            join(workspacePath, "foo", "bar", "baz"),
+            "foo / bar / baz",
+            env,
+            [
+              new FileTreeLeaf(
+                join(workspacePath, "foo", "bar", "baz", "query2.ql"),
+                "query2.ql",
+                "java",
+              ),
+            ],
+          ),
+          new FileTreeLeaf(
+            join(workspacePath, "query1.ql"),
+            "query1.ql",
+            "java",
+          ),
+        ]),
+      ]);
+    });
+  });
 
-      expect(onDidChangeQueriesSpy).toHaveBeenCalledTimes(2);
+  describe("recomputeAllQueryLanguages", () => {
+    it("should recompute the language of all query files", async () => {
+      makeTestFile(join(workspacePath, "query.ql"));
+
+      await discovery.initialRefresh();
+
+      expect(discovery.buildQueryTree()).toEqual([
+        new FileTreeDirectory(workspacePath, "workspace", env, [
+          new FileTreeLeaf(join(workspacePath, "query.ql"), "query.ql", "java"),
+        ]),
+      ]);
+
+      queryPackDiscoverer.getLanguageForQueryFile = () => QueryLanguage.Python;
+      onDidChangeQueryPacks.fire();
+
+      // Wait for the query discovery to recompute the query languages.
+      // This is async but should complete instantly since it's all in-memory.
+      await sleep(100);
```

Comment:
Sadly the language re-computation is now async, but it should complete instantly since it's just calling the mocked `getLanguageForQueryFile` method in a loop. I think this sleep should be pretty safe and not be a risk of flakiness. I'd love if we could avoid this sleep, but I don't think we can right now.